probe-rs/src/core.rs

```diff
@@ -163,6 +174,244 @@ pub trait CoreInterface: MemoryInterface {
+        Ok(self.registers[&register])
+    }
+
+    fn id(&self) -> usize {
+        todo!()
+    }
+
+    fn wait_for_core_halted(&mut self, _timeout: std::time::Duration) -> Result<(), error::Error> {
+        todo!()
+    }
+
+    fn core_halted(&mut self) -> Result<bool, error::Error> {
+        todo!()
+    }
+
+    fn status(&mut self) -> Result<crate::CoreStatus, error::Error> {
+        todo!()
+    }
+
+    fn halt(
+        &mut self,
+        _timeout: std::time::Duration,
+    ) -> Result<crate::CoreInformation, error::Error> {
+        todo!()
+    }
+
+    fn run(&mut self) -> Result<(), error::Error> {
+        todo!()
+    }
+
+    fn reset(&mut self) -> Result<(), error::Error> {
+        todo!()
+    }
+
+    fn reset_and_halt(
+        &mut self,
+        _timeout: std::time::Duration,
+    ) -> Result<crate::CoreInformation, error::Error> {
+        todo!()
+    }
+
+    fn step(&mut self) -> Result<crate::CoreInformation, error::Error> {
+        todo!()
+    }
+
+    fn write_core_reg(
+        &mut self,
+        _address: crate::core::registers::RegisterId,
+        _value: crate::core::registers::RegisterValue,
+    ) -> Result<(), error::Error> {
+        todo!()
+    }
+
+    fn available_breakpoint_units(&mut self) -> Result<u32, error::Error> {
+        todo!()
+    }
+
+    fn hw_breakpoints(&mut self) -> Result<Vec<Option<u64>>, error::Error> {
+        todo!()
+    }
+
+    fn enable_breakpoints(&mut self, _state: bool) -> Result<(), error::Error> {
+        todo!()
+    }
+
+    fn set_hw_breakpoint(&mut self, _unit_index: usize, _addrr: u64) -> Result<(), error::Error> {
+        todo!()
+    }
+
+    fn clear_hw_breakpoint(&mut self, _unit_index: usize) -> Result<(), error::Error> {
+        todo!()
+    }
+
+    fn registers(&self) -> &'static crate::core::registers::CoreRegisters {
+        // TODO: determine correct registers
+        &CORTEX_M_CORE_REGISTERS
+    }
+
+    fn program_counter(&self) -> &'static CoreRegister {
+        todo!()
+    }
+
+    fn frame_pointer(&self) -> &'static CoreRegister {
+        todo!()
+    }
+
+    fn stack_pointer(&self) -> &'static CoreRegister {
+        todo!()
+    }
+
+    fn return_address(&self) -> &'static CoreRegister {
+        todo!()
+    }
+
+    fn hw_breakpoints_enabled(&self) -> bool {
+        todo!()
+    }
+
+    fn architecture(&self) -> probe_rs_target::Architecture {
+        todo!()
+    }
+
+    fn core_type(&self) -> probe_rs_target::CoreType {
+        // TODO: determine correct type
+        probe_rs_target::CoreType::Armv6m
+    }
+
+    fn fpu_support(&mut self) -> Result<bool, error::Error> {
+        todo!()
+    }
+
+    fn reset_catch_set(&mut self) -> Result<(), crate::Error> {
+        todo!()
+    }
+
+    fn reset_catch_clear(&mut self) -> Result<(), crate::Error> {
+        todo!()
+    }
+
+    fn debug_core_stop(&mut self) -> Result<(), crate::Error> {
+        todo!()
+    }
+}
```

Comment:
Is this necessary for the core dump? Wouldn't it be enough to just have the `MemoryInterface`?

Reply:
No, it has to figure out the architecture and some registers.

Reply:
I remodelled it initially but then decided it is cleaner this way.

Reply:
I don't like this because you have lots of functions will will never work. And if you pass this trait everywhere, then you never know which function actually does something with the core, and which ones just need access to some memory.

Reply:
Well, I am happy to iterate this, but I think it's not necessarily feasible to remodel even more in this PR. Wdyt?

Reply:
Sure, we can use the argument to end all reviews and just merge it because it's too big :)

Reply:
Done.